ccl/sqlproxyccl/tenant: TestWatchPods failed

ccl/sqlproxyccl/tenant.TestWatchPods [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsIbmCloudLinuxS390x_UnitTestsS390x/20895558?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsIbmCloudLinuxS390x_UnitTestsS390x/20895558?buildTab=artifacts#/) on master @ [3c84fdbda2020373e08a6456c6f07fdd8a285e68](https://github.com/cockroachdb/cockroach/commits/3c84fdbda2020373e08a6456c6f07fdd8a285e68):


```
=== RUN   TestWatchPods
    test_log_scope.go:171: test logs captured to: /artifacts/tmp/_tmp/7cf74e9036d0a0f6afa915d836c9678f/logTestWatchPods1369044360
    test_log_scope.go:82: use -show-logs to present logs inline
    directory_cache_test.go:317: 
        	Error Trace:	pkg/ccl/sqlproxyccl/tenant/directory_cache_test.go:317
        	Error:      	Not equal: 
        	            	expected: 2
        	            	actual  : 1
        	Test:       	TestWatchPods
    panic.go:615: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/7cf74e9036d0a0f6afa915d836c9678f/logTestWatchPods1369044360
--- FAIL: TestWatchPods (1.03s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestWatchPods.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-58117

--- a/pkg/ccl/sqlproxyccl/tenantdirsvr/test_static_directory_svr.go
+++ b/pkg/ccl/sqlproxyccl/tenantdirsvr/test_static_directory_svr.go
@@ -135,11 +135,6 @@ func (d *TestStaticDirectoryServer) WatchPods(
 		return status.Errorf(codes.FailedPrecondition, "directory server has not been started")
 	}
 
-	addListener := func(ch chan *tenant.WatchPodsResponse) *list.Element {
-		d.mu.Lock()
-		defer d.mu.Unlock()
-		return d.mu.podEventListeners.PushBack(ch)
-	}
 	removeListener := func(e *list.Element) chan *tenant.WatchPodsResponse {
 		d.mu.Lock()
 		defer d.mu.Unlock()
@@ -158,11 +153,17 @@ func (d *TestStaticDirectoryServer) WatchPods(
 	// Construct the channel with a small buffer to allow for a burst of
 	// notifications, and a slow receiver.
 	c := make(chan *tenant.WatchPodsResponse, 10)
-	chElement := addListener(c)
 
+	// Capture the initial pods and add the listener atomically within the same
+	// lock hold. This prevents a race where a pod update occurs after the
+	// listener is added but before we capture the initial state, which would
+	// cause duplicate events (the update notification and the initial snapshot
+	// would both contain the same pod).
+	var chElement *list.Element
 	initialPods := func() (result []tenant.WatchPodsResponse) {
 		d.mu.Lock()
 		defer d.mu.Unlock()
+		chElement = d.mu.podEventListeners.PushBack(c)
 		for _, pods := range d.mu.tenantPods {
 			for _, pod := range pods {
 				result = append(result, tenant.WatchPodsResponse{
@@ -265,11 +266,6 @@ func (d *TestStaticDirectoryServer) WatchTenants(
 		return status.Errorf(codes.FailedPrecondition, "directory server has not been started")
 	}
 
-	addListener := func(ch chan *tenant.WatchTenantsResponse) *list.Element {
-		d.mu.Lock()
-		defer d.mu.Unlock()
-		return d.mu.tenantEventListeners.PushBack(ch)
-	}
 	removeListener := func(e *list.Element) chan *tenant.WatchTenantsResponse {
 		d.mu.Lock()
 		defer d.mu.Unlock()
@@ -288,11 +284,16 @@ func (d *TestStaticDirectoryServer) WatchTenants(
 	// Construct the channel with a small buffer to allow for a burst of
 	// notifications, and a slow receiver.
 	c := make(chan *tenant.WatchTenantsResponse, 10)
-	chElement := addListener(c)
 
+	// Capture the initial tenants and add the listener atomically within the
+	// same lock hold. This prevents a race where a tenant update occurs after
+	// the listener is added but before we capture the initial state, which
+	// would cause duplicate events.
+	var chElement *list.Element
 	initialTenants := func() (result []tenant.WatchTenantsResponse) {
 		d.mu.Lock()
 		defer d.mu.Unlock()
+		chElement = d.mu.tenantEventListeners.PushBack(c)
 		for id := range d.mu.tenants {
 			result = append(result, tenant.WatchTenantsResponse{
 				Type:   tenant.EVENT_ADDED,

